lib/linked_list.rb

```diff
@@ -12,120 +11,264 @@ def initialize(value, next_node = nil)
-    def has_cycle
-      raise NotImplementedError
+    return min
+  end
+  
+  # method that returns the length of the singly linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def length    
+    current = @head
+    length = 0
+    
+    until current.nil?
+      length += 1
+      current = current.next  
     end
-
-
-    # Additional Exercises 
-    # returns the value in the first node
-    # returns nil if the list is empty
-    def get_first
-      raise NotImplementedError
+    return length
+  end
+  
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def get_at_index(index)
+    current = @head
+    count = 0 
+    
+    until current.nil?
+      if (count == index)
+        return current.data
+      end
+      count += 1
+      current = current.next
     end
-
-    # method that inserts a given value as a new last node in the linked list
-    def add_last(value)
-      raise NotImplementedError
+  end
+  
+  # method to print all the values in the linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def visit    
+    current = @head
+    
+    until current.nil?
+      print current.data
+      current = current.next
     end
-
-    # method that returns the value of the last node in the linked list
-    # returns nil if the linked list is empty
-    def get_last
-      raise NotImplementedError
+  end
+  
+  # method to delete the first node found with specified value
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def delete(value)
+    current = @head
+    previous = nil
+    
+    until current.nil?
+      if current.data == value
+        if previous.nil?
+          @head = current.next
+          return current.next = current.next.next
+        else
+          previous.next = current.next
+          break
+        end 
+      end 
+      
+      previous = current
+      current = current.next
     end
-
-    # method to insert a new node with specific data value, assuming the linked
-    # list is sorted in ascending order
-    def insert_ascending(value)
-      raise NotImplementedError
+  end 
+  
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  # Time Complexity: O(n)
+  # Space Complexity: O(1) - reverse in place
+  def reverse
+    return if @head.nil?
+    
+    current = @head
+    previous = nil
+    
+    until current.next.nil?
+      temp = current.next
+      current.next = previous
+      previous = current
+      current = temp
+    end 
+    
+    current.next = previous
+    @head = current
+  end
+  
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity:
+  # Space Complexity:
+  def find_middle_value
+    raise NotImplementedError
+  end

```

Comment:
Consider having two references, slow and fast.  Fast moves two nodes for every node slow moves to.  You loop until `fast` is `nil` and then slow will be the middle node.